### ISSUE TITLE
Could not find Java_java_lang_ClassLoader_00024NativeLibrary_load

### ISSUE THREAD

**Issue by holy-moly-555** [SAP/SapMachine#1238]
Hey there,

we recently updated our SapMachine from 11.0.14 to 11.0.16. We use the latest Elastic APM Java Agent with the profiling feature (it uses the async-profiler project). After the update the following warnings occurr frequently in the logs:

`WARNING: Could not find Java_java_lang_ClassLoader_00024NativeLibrary_load`

It seems like the issue came with changes in openJDK11. See these post for reference:

- https://stackoverflow.com/questions/72178216/java-error-java-java-lang-classloader-00024nativelibrary-load0
- https://github.com/openjdk/jdk11u/commit/f7346f087a14f1878029a00104005e04468d333f#diff-c106c06bd7c092b3b91cb0487c089c1e96d33305c6bf599e2082a709271e3948R338

Do you plan to fix this issue?

Kind regards.

**Comment by RealCLanger** [comment]
Hi,

Seems like there is an unresolved OpenJDK bug about this: https://bugs.openjdk.org/browse/JDK-8288547

Is it possible to create a reproducer? Then we could have a look.

Best regards
Christoph

**Comment by MBaesken** [comment]
A reproducer would be a (hopefully small) program showing the issue. For example a Java program or shell script or something that can be used to show that it 'worked' with 11.0.14 and changed  runtime behavior with 11.0.15 .
This could be added to the existing JBS issue https://bugs.openjdk.org/browse/JDK-8288547 .

**Comment by holy-moly-555** [comment]
Thanks for the clarification @MBaesken .
Do you think it's safe to still run the apm agent alongside our application altough the warning is logged? Can you assess the impact on the jvm due to this bug?

**Comment by MBaesken** [comment]
As long as the APM agent is just showing the warning why not continue running it if it brings you a benefit.
Regarding impact on JVM, the symbol name was load0 for a long time.
When looking into a very old SapMachine 11 from ~ 3 years ago, even back then the symbol name was *load0 not *load 
nm /OpenSource/sapmachine/11/lib/libjava.so | grep NativeLibrary
000000000000e3f0 T Java_java_lang_ClassLoader_00024NativeLibrary_findEntry
000000000000deb0 T Java_java_lang_ClassLoader_00024NativeLibrary_load0
000000000000e250 T Java_java_lang_ClassLoader_00024NativeLibrary_unload

However back in old jdk8 times I think the symbol name was  Java_java_lang_ClassLoader_00024NativeLibrary_load but this changed years ago to load0.

**Comment by tstuefe** [comment]
I find it highly likely that async-profiler, part of the epm package, is at fault here.

I ran the reproduction scenario (https://github.com/santipyx/oracle-9170972) given in the associated message.

Could reproduce the messages.

Removing the java agent removed the messages.

Async profiler does - or did at some point, not sure if they still do - resolve NativeLibrary::load resp. load0 via reflection. See https://github.com/jvm-profiling-tools/async-profiler/blob/5312a793ec22106420883e8a274d10c390e6e4b1/src/profiler.cpp#L582 .

Not sure if that was a bug that was fixed, and the epm package is just outdated.

**Comment by tstuefe** [comment]
Correction, they don't try to resolve it, they try to intercept it via RegisterNatives :-(

**Comment by MBaesken** [comment]
looks like the method signature change to  NativeLibrary  load0 in 11.0.15 (one boolean parameter more) does not work well together with the checks here 
https://github.com/jvm-profiling-tools/async-profiler/blob/5312a793ec22106420883e8a274d10c390e6e4b1/src/profiler.cpp#L590
So async-profiler needs a bit of updating .

**Comment by RealCLanger** [comment]
Hi,

I did some further digging and found out that the problematic code in Async Profiler was removed earlier this year by this commit: https://github.com/jvm-profiling-tools/async-profiler/commit/3456dd3d9040ede77fb0350cba8b63cd1fc3c5e2

However, when I'm looking at [apm-agent](https://github.com/elastic/apm-agent-java), I can see that async profiler version 1.8.7 is used in the latest version or even 1.8.3, when I look [here](https://github.com/elastic/apm-agent-java/blob/main/apm-agent-plugins/apm-profiling-plugin/pom.xml). The async change, however, is probably part of some 2.* version.

So I suggest to report this issue at [apm-agent](https://github.com/elastic/apm-agent-java) and see if they can fix this. 

With this, I'll close this SapMachine issue report.

Thanks
Christoph

@apangin for your attention.

**Comment by holy-moly-555** [comment]
Thanks for investigating.

**Comment by apangin** [comment]
@RealCLanger You're right - async-profiler tried to intercept newly loaded libraries by hooking JNI implementation of `NativeLibrary.load`. Unfortunately, there is no standard supported API to do that.

Some time ago, as I found the signatures of NativeLibrary methods quite unstable (they changed back and forth even in minor JDK updates), I finally gave up playing version guessing game and completely rewrote the hooking mechanism. Since version 2.7, async-profiler instruments libc `dlopen` instead.
